[activation checkpointing] Higher order functional rng op wrappers

anijain2305 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #102935
* #103357
* __->__ #102934



Introduces two higher order operators
* run_and_save_rng_state - Saves the current rng state and then runs the op.
* run_with_rng_state - Runs the op with the rng state supplied as an input

Ideally, we would like to use torch.compile for these operators. But currently the plan is to introduce these operators at the partitioner level, obviating the need to support them fully through the torch.compile stack. To ensure that we have good enough debugging with minifiers, we have ensure that they work with make_fx. In future, we can move on torch.compile.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @aakhundov

anijain2305 committed:
[activation checkpointing] Higher order functional rng op wrappers

[ghstack-poisoned]

anijain2305 committed:
Update on "[activation checkpointing] Higher order functional rng op wrappers"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225

[ghstack-poisoned]

anijain2305 committed:
Update on "[activation checkpointing] Higher order functional rng op wrappers"




Introduces two higher order operators
* run_and_save_rng_state - Saves the current rng state and then runs the op.
* run_with_rng_state - Runs the op with the rng state supplied as an input

Together these two ops can be used to functionalize the rng operators in the forward and recomputed forward in the backward pass (for activation checkpointing). Currently, the plan to introduce these operators is at the partitioner level, so they don't need to be supported via Dynamo, AOT AUtograd stack. But, there is just enough implementation to survive `make_fx` in order to work with minifier.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225

[ghstack-poisoned]

anijain2305 committed:
Update on "[activation checkpointing] Higher order functional rng op wrappers"




Introduces two higher order operators
* run_and_save_rng_state - Saves the current rng state and then runs the op.
* run_with_rng_state - Runs the op with the rng state supplied as an input

Together these two ops can be used to functionalize the rng operators in the forward and recomputed forward in the backward pass (for activation checkpointing). Currently, the plan to introduce these operators is at the partitioner level, so they don't need to be supported via Dynamo, AOT AUtograd stack. But, there is just enough implementation to survive `make_fx` in order to work with minifier.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225

[ghstack-poisoned]

anijain2305 committed:
Update on "[activation checkpointing] Higher order functional rng op wrappers"




Introduces two higher order operators
* run_and_save_rng_state - Saves the current rng state and then runs the op.
* run_with_rng_state - Runs the op with the rng state supplied as an input

Ideally, we would like to use torch.compile for these operators. But currently the plan is to introduce these operators at the partitioner level, obviating the need to support them fully through the torch.compile stack. To ensure that we have good enough debugging with minifiers, we have ensure that they work with make_fx. In future, we can move on torch.compile.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225

[ghstack-poisoned]

anijain2305 committed:
Update on "[activation checkpointing] Higher order functional rng op wrappers"




Introduces two higher order operators
* run_and_save_rng_state - Saves the current rng state and then runs the op.
* run_with_rng_state - Runs the op with the rng state supplied as an input

Ideally, we would like to use torch.compile for these operators. But currently the plan is to introduce these operators at the partitioner level, obviating the need to support them fully through the torch.compile stack. To ensure that we have good enough debugging with minifiers, we have ensure that they work with make_fx. In future, we can move on torch.compile.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225

[ghstack-poisoned]

zou3519 commented on torch/_ops.py:
```diff
@@ -251,8 +251,11 @@ def __call__(self, *args, **kwargs):
         dispatch_key_set = _compute_keyset(args, kwargs, self.non_fallthrough_keys)
         return self.dispatch(dispatch_key_set.highestPriorityTypeId(), *args, **kwargs)
 
+    def __str__(self):
+        return f"torch.ops.{self.name()}"
+
```
Should just be `{self.name()}`. It's not guaranteed that a HigherOrderOp is accessible under torch.ops

zou3519 commented on torch/_prims/rng_prims.py:
```diff
@@ -132,5 +148,163 @@ def _philox_rand(
+                raise NotImplementedError(
+                    f"Backend not available for device {device_type}"
+                )
+        else:
```
Assuming there is a Tensor arg, we just re-dispatch and the dispatcher will take care of it for us:
```
with _ExcludeDispatchKeyGuard(DispatchKeySet(DispatchKey.BackendSelect)):
    run_and_save_rng_state(*args, **kwargs)
```

anijain2305 replied:
I tried that earlier, but I was getting lots of failures. 

As of now, I get this error, when I use the above suggestion. It is going through my fake tensor impl, so something might be bad there?

~~~
  File "/scratch/anijain/work/pytorch/torch/_ops.py", line 252, in __call__
    return self.dispatch(dispatch_key_set.highestPriorityTypeId(), *args, **kwargs)
  File "/scratch/anijain/work/pytorch/torch/_ops.py", line 230, in dispatch
    return self.python_key_mode_table[type(curr_mode)](*args, **kwargs)
  File "/scratch/anijain/work/pytorch/torch/_prims/rng_prims.py", line 202, in impl_fake_tensor_mode
    return impl_backend_select(op, *args, **kwargs)
  File "/scratch/anijain/work/pytorch/torch/_prims/rng_prims.py", line 186, in impl_backend_select
    return run_and_save_rng_state(op, *args, **kwargs)
  File "/scratch/anijain/work/pytorch/torch/_ops.py", line 252, in __call__
    return self.dispatch(dispatch_key_set.highestPriorityTypeId(), *args, **kwargs)
  File "/scratch/anijain/work/pytorch/torch/_ops.py", line 230, in dispatch
    return self.python_key_mode_table[type(curr_mode)](*args, **kwargs)
  File "/scratch/anijain/work/pytorch/torch/_prims/rng_prims.py", line 202, in impl_fake_tensor_mode
    return impl_backend_select(op, *args, **kwargs)
  File "/scratch/anijain/work/pytorch/torch/_prims/rng_prims.py", line 186, in impl_backend_select
    return run_and_save_rng_state(op, *args, **kwargs)
  File "/scratch/anijain/work/pytorch/torch/_ops.py", line 245, in __call__
    flat_args = _to_flat_tuple(args, kwargs)
  File "/scratch/anijain/work/pytorch/torch/_ops.py", line 262, in _to_flat_tuple
    flat_args, _ = torch.utils._pytree.tree_flatten(args)
  File "/scratch/anijain/work/pytorch/torch/utils/_pytree.py", line 274, in tree_flatten
    if _is_leaf(pytree):
  File "/scratch/anijain/work/pytorch/torch/utils/_pytree.py", line 233, in _is_leaf
    return _get_node_type(pytree) not in SUPPORTED_NODES
  File "/scratch/anijain/work/pytorch/torch/utils/_pytree.py", line 227, in _get_node_type
    if _is_namedtuple_instance(pytree):
  File "/scratch/anijain/work/pytorch/torch/utils/_pytree.py", line 217, in _is_namedtuple_instance
    typ = type(pytree)
torch._dynamo.exc.BackendCompilerFailed: backend='inductor' raised:
RecursionError: maximum recursion depth exceeded while calling a Python object

While executing %run_and_save_rng_state : [#users=2] = call_function[target=torch.ops.run_and_save_rng_state](args = (aten.native_dropout.default, %div_1, 0.1, True), kwargs = {})
Original traceback:
None

~~~

zou3519 replied:
Let me poke a bit at this locally and get back to you

zou3519 replied:
What is written in this function is not entirely faithful to what the PyTorch dispatcher does for regular operators, but I'm happy to accept this as-is to unblock.

The problem is that the PyTorch dispatcher uses "redispatch" to implement BackendSelect. What "redispatch" would do here is that for subsequent invocations of this operator, it masks out the dispatch keys above and including BackendSelect so that a dispatch key below it can be selected. Compare this to an ExcludeDispatchKeyGuard, which turns off the dispatch key for all operators (not just this operator!), which ends up causing problems down the line.

We have not implemented a "redispatch" for HigherOrderOperator yet.

anijain2305 replied:
I see, that makes sense. I am totally willing to reimplement this function once we have redispatch for HigherOrderOp. So, lets go w/ the current design.

zou3519 commented on torch/_prims/rng_prims.py:
```diff
@@ -132,5 +148,163 @@ def _philox_rand(
+
+
+def register_run_with_rng_state_op():
+    run_with_rng_state = HigherOrderOperator("run_with_rng_state")
+
```
How do you access `run_with_rng_state` later? I don't see it being returned from this function

anijain2305 replied:
Do we have to return? 

`HigherOrderOperator("run_with_rng_state")` seems to register a `torch.ops.run_with_rng_state` already. So, I don't really need to return this.

It could totally be that my understanding is incorrect. So, please correct me if I am wrong.

zou3519 replied:
You're right, I didn't realize this is what the code did.

zou3519 commented on test/test_prims.py:
```diff
@@ -1124,6 +1124,31 @@ def test_philox_rand(self, device, dtype):
                 self.assertEqual(a, b)
 
+
+    @onlyCUDA
+    @dtypes(torch.float32)
+    def test_functional_rng_wrappers(self, device, dtype):
```
Should this also test CPU?

zou3519 commented on torch/_prims/rng_prims.py:
```diff
@@ -132,5 +148,163 @@ def _philox_rand(
+
+    @run_with_rng_state.py_impl(DispatchKey.BackendSelect)
+    def impl_backend_select(rng_state, op, *args, **kwargs):
+        # with _ExcludeDispatchKeyGuard(DispatchKeySet(DispatchKey.BackendSelect)):
```
delete unused code?

anijain2305 committed:
Update on "[activation checkpointing] Higher order functional rng op wrappers"




Introduces two higher order operators
* run_and_save_rng_state - Saves the current rng state and then runs the op.
* run_with_rng_state - Runs the op with the rng state supplied as an input

Ideally, we would like to use torch.compile for these operators. But currently the plan is to introduce these operators at the partitioner level, obviating the need to support them fully through the torch.compile stack. To ensure that we have good enough debugging with minifiers, we have ensure that they work with make_fx. In future, we can move on torch.compile.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225

[ghstack-poisoned]

jansel commented on torch/_prims/rng_prims.py:
```diff
@@ -132,5 +148,161 @@ def _philox_rand(
+        if kwargs.get("device"):
+            device = kwargs.get("device")
+            device_type = device if isinstance(device, str) else device.type  # type: ignore[union-attr]
+            if "cuda" in device_type:
```
why `in` vs `==`

jansel commented on torch/_prims/rng_prims.py:
```diff
@@ -132,5 +148,161 @@ def _philox_rand(
+            device_type = device if isinstance(device, str) else device.type  # type: ignore[union-attr]
+            if "cuda" in device_type:
+                impl = impl_cuda
+            elif "cpu" in device_type:
```
why `in` vs `==`

anijain2305 commented on test/inductor/test_torchinductor.py:
```diff
@@ -4799,6 +4799,50 @@ def fn(a):
+        # minifiers, we have ensure that they work with make_fx. This test uses
+        # make_fx to do the testing. In future, we can move on torch.compile.
+        def fn():
+            rng_state1, a1 = torch.ops.run_and_save_rng_state(
```
The ones in aten/prims do not accept a torch.ops.OpOverload as an operand. It was unclear what the schema for such ops should look like. So, higher order op seemed a better fit to me. What do you think?

zou3519 replied:
This is a pre-existing condition. All of the HigherOrderOps are in the global namespace (which is problematic for other things), we should fix it at some point

zou3519 replied:
@anijain2305 My plan is to stop exposing HigherOrderOperators under torch.ops.*. The main reasons are:
1) redundancy. We're going to have e.g. torch.cond, having it also as torch.ops.cond is redundant. Comparing this to the rest of torch.ops, having aten operators like torch.ops.aten.sin is reasonable because there is no other way to access them from Python.
2) Being in a global namespace leads to problems

The implication is that we should expose `run_and_save_rng_state` as e.g. torch._prims.rng_prims.run_and_save_rng_state and access them from there directly instead of under torch.ops

anijain2305 replied:
Yes, that makes sense. Let me change the code to access the ops.

anijain2305 commented on torch/_prims/rng_prims.py:
```diff
@@ -132,5 +148,161 @@ def _philox_rand(
+    run_and_save_rng_state.fallthrough(DispatchKey.PythonDispatcher)  # type: ignore[attr-defined]
+    run_and_save_rng_state.fallthrough(DispatchKey.PythonTLSSnapshot)  # type: ignore[attr-defined]
+
+    @run_and_save_rng_state.py_impl(DispatchKey.Autograd)
+    def impl_autograd(op, *args, **kwargs):
+        with torch._C._AutoDispatchBelowAutograd():
+            return run_and_save_rng_state(op, *args, **kwargs)
```
Yes, not now. We are introducing these in partitioners for now. But, I will be very happy to pick this task up in a few weeks once Activation checkpointing is in good state.

anijain2305 committed:
Update on "[activation checkpointing] Higher order functional rng op wrappers"




Introduces two higher order operators
* run_and_save_rng_state - Saves the current rng state and then runs the op.
* run_with_rng_state - Runs the op with the rng state supplied as an input

Ideally, we would like to use torch.compile for these operators. But currently the plan is to introduce these operators at the partitioner level, obviating the need to support them fully through the torch.compile stack. To ensure that we have good enough debugging with minifiers, we have ensure that they work with make_fx. In future, we can move on torch.compile.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225

[ghstack-poisoned]

anijain2305 committed:
Update on "[activation checkpointing] Higher order functional rng op wrappers"




Introduces two higher order operators
* run_and_save_rng_state - Saves the current rng state and then runs the op.
* run_with_rng_state - Runs the op with the rng state supplied as an input

Ideally, we would like to use torch.compile for these operators. But currently the plan is to introduce these operators at the partitioner level, obviating the need to support them fully through the torch.compile stack. To ensure that we have good enough debugging with minifiers, we have ensure that they work with make_fx. In future, we can move on torch.compile.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225

[ghstack-poisoned]

zou3519 commented on torch/_inductor/ir.py:
```diff
@@ -3225,6 +3225,8 @@ def __init__(
                 if V.graph.cpp_wrapper
                 else f"aten.{kernel.__name__}"
             )
+        elif getattr(torch.ops, kernel.__name__, None) is kernel:
+            self.kernel = f"torch.ops.{kernel.__name__}"
```
Does this hardcode torch.ops.run_and_save_rng_state ? (can we avoid this?)

anijain2305 replied:
Ah good catch. Fixed it now.

zou3519 commented on torch/_inductor/ir.py:
```diff
@@ -3283,6 +3285,24 @@ def __repr__(self):
         return args
 
+    @staticmethod
+    def find_device(tensor_args, example_output):
```
Just for my understanding, what is this used for?

anijain2305 replied:
We use fallback for few kernels in TorchInductor. For those kernels, we look at the inputs and outputs of the op to figure out the device, so that we can appropriately create the TorchInductor IR node.

zou3519 reviewed (commented):
Looks like there's still one instance of torch.ops.* baked in, but other than that, this looks fine to me

anijain2305 committed:
Update on "[activation checkpointing] Higher order functional rng op wrappers"




Introduces two higher order operators
* run_and_save_rng_state - Saves the current rng state and then runs the op.
* run_with_rng_state - Runs the op with the rng state supplied as an input

Ideally, we would like to use torch.compile for these operators. But currently the plan is to introduce these operators at the partitioner level, obviating the need to support them fully through the torch.compile stack. To ensure that we have good enough debugging with minifiers, we have ensure that they work with make_fx. In future, we can move on torch.compile.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225

[ghstack-poisoned]

anijain2305 commented:
@jansel This is ready for review. PTAL.

anijain2305 committed:
Update on "[activation checkpointing] Higher order functional rng op wrappers"




Introduces two higher order operators
* run_and_save_rng_state - Saves the current rng state and then runs the op.
* run_with_rng_state - Runs the op with the rng state supplied as an input

Ideally, we would like to use torch.compile for these operators. But currently the plan is to introduce these operators at the partitioner level, obviating the need to support them fully through the torch.compile stack. To ensure that we have good enough debugging with minifiers, we have ensure that they work with make_fx. In future, we can move on torch.compile.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225

[ghstack-poisoned]

anijain2305 committed:
Update on "[activation checkpointing] Higher order functional rng op wrappers"




Introduces two higher order operators
* run_and_save_rng_state - Saves the current rng state and then runs the op.
* run_with_rng_state - Runs the op with the rng state supplied as an input

Ideally, we would like to use torch.compile for these operators. But currently the plan is to introduce these operators at the partitioner level, obviating the need to support them fully through the torch.compile stack. To ensure that we have good enough debugging with minifiers, we have ensure that they work with make_fx. In future, we can move on torch.compile.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225

[ghstack-poisoned]

anijain2305 committed:
Update on "[activation checkpointing] Higher order functional rng op wrappers"




Introduces two higher order operators
* run_and_save_rng_state - Saves the current rng state and then runs the op.
* run_with_rng_state - Runs the op with the rng state supplied as an input

Ideally, we would like to use torch.compile for these operators. But currently the plan is to introduce these operators at the partitioner level, obviating the need to support them fully through the torch.compile stack. To ensure that we have good enough debugging with minifiers, we have ensure that they work with make_fx. In future, we can move on torch.compile.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225

[ghstack-poisoned]

anijain2305 committed:
Update on "[activation checkpointing] Higher order functional rng op wrappers"




Introduces two higher order operators
* run_and_save_rng_state - Saves the current rng state and then runs the op.
* run_with_rng_state - Runs the op with the rng state supplied as an input

Ideally, we would like to use torch.compile for these operators. But currently the plan is to introduce these operators at the partitioner level, obviating the need to support them fully through the torch.compile stack. To ensure that we have good enough debugging with minifiers, we have ensure that they work with make_fx. In future, we can move on torch.compile.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225

[ghstack-poisoned]

anijain2305 committed:
Update on "[activation checkpointing] Higher order functional rng op wrappers"




Introduces two higher order operators
* run_and_save_rng_state - Saves the current rng state and then runs the op.
* run_with_rng_state - Runs the op with the rng state supplied as an input

Ideally, we would like to use torch.compile for these operators. But currently the plan is to introduce these operators at the partitioner level, obviating the need to support them fully through the torch.compile stack. To ensure that we have good enough debugging with minifiers, we have ensure that they work with make_fx. In future, we can move on torch.compile.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225

[ghstack-poisoned]

zou3519 reviewed (approved):
Higher-order operator pieces LGTM, you might want someone else to look over the inductor pieces.

jansel commented on test/test_prims.py:
```diff
@@ -1124,6 +1124,28 @@ def test_philox_rand(self, device, dtype):
+        rng_state1, res1 = torch._prims.rng_prims.run_and_save_rng_state(torch.rand, 10, device=device, dtype=dtype)
+        rng_state2, res2 = torch._prims.rng_prims.run_and_save_rng_state(torch.rand, 10, device=device, dtype=dtype)
+
+        torch.manual_seed(123)
```
Shouldn't this not be needed?

anijain2305 replied:
Ah. Good catch! yes, this is not needed. Fixing.

anijain2305 committed:
Update on "[activation checkpointing] Higher order functional rng op wrappers"




Introduces two higher order operators
* run_and_save_rng_state - Saves the current rng state and then runs the op.
* run_with_rng_state - Runs the op with the rng state supplied as an input

Ideally, we would like to use torch.compile for these operators. But currently the plan is to introduce these operators at the partitioner level, obviating the need to support them fully through the torch.compile stack. To ensure that we have good enough debugging with minifiers, we have ensure that they work with make_fx. In future, we can move on torch.compile.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225

[ghstack-poisoned]

anijain2305 commented:
@pytorchbot merge
